tests/import_specifications/test_individual_parsers.py

```diff
@@ -0,0 +1,235 @@
+        ErrorType.PARSE_FAIL, message, SpecificationSource(input_)
+    )]))
+    assert res == expected
+
+
+def test_xsv_parse_fail_empty_file(temp_dir: Path):
+    _xsv_parse_fail(temp_dir, [], parse_csv, "Missing data type / version header")
+
+
+def test_xsv_parse_fail_bad_datatype_header(temp_dir: Path):
+    err = ('Invalid header; got "This is the wrong header", expected "Data type: '
+        + '<data_type>; Version: <version>"')
+    _xsv_parse_fail(temp_dir, ["This is the wrong header"], parse_csv, err)
+
+
+def test_xsv_parse_fail_bad_version(temp_dir: Path):
+    err = "Schema version 87 is larger than maximum processable version 1"
+    _xsv_parse_fail(temp_dir, ["Data type: foo; Version: 87"], parse_csv, err)
+
+def test_xsv_parse_fail_missing_column_headers(temp_dir: Path):
+    err = "Expected 2 column header rows"
+    _xsv_parse_fail(temp_dir, ["Data type: foo; Version: 1\n"], parse_csv, err)
+
+    lines = ["Data type: foo; Version: 1\n", "head1, head2\n"]
+    _xsv_parse_fail(temp_dir, lines, parse_csv, err)
+
+
+def test_xsv_parse_fail_missing_data(temp_dir: Path):
+    err = "No data in file"
+    lines = [
+        "Data type: foo; Version: 1\n"
+        "head1, head2, head3\n",
+        "Head 1, Head 2, Head 3\n",
+    ]
+    _xsv_parse_fail(temp_dir, lines, parse_csv, err)
+
+
+def test_xsv_parse_fail_unequal_rows(temp_dir: Path):
+    err = "Header rows have unequal column counts"
+    lines = [
+        "Data type: foo; Version: 1\n"
+        "head1, head2, head3\n",
+        "Head 1, Head 2\n",
+    ]
+    _xsv_parse_fail(temp_dir, lines, parse_csv, err)
+
+    err = "Incorrect number of items in line 3, expected 2, got 3"
+    lines = [
+        "Data type: foo; Version: 1\n"
+        "head1\thead2\n",
+        "Head 1\tHead 2\tHead 3\n",
+    ]
+    _xsv_parse_fail(temp_dir, lines, parse_tsv, err)
+
+    err = "Incorrect number of items in line 5, expected 3, got 4"
+    lines = [
+        "Data type: foo; Version: 1\n"
+        "head1, head2, head 3\n",
+        "Head 1, Head 2, Head 3\n",
+        "1, 2, 3\n",
+        "1, 2, 3, 4\n",
+        "1, 2, 3\n",
+    ]
+    _xsv_parse_fail(temp_dir, lines, parse_csv, err)
+
+    err = "Incorrect number of items in line 6, expected 3, got 2"
+    lines = [
+        "Data type: foo; Version: 1\n"
+        "head1\thead2\thead 3\n",
+        "Head 1\tHead 2\tHead 3\n",
+        "1\t2\t3\n",
+        "1\t2\t3\n",
+        "1\t2\n",
+        "1\t2\t3\n",
+    ]
+    _xsv_parse_fail(temp_dir, lines, parse_tsv, err)
+
+    err = "Incorrect number of items in line 5, expected 3, got 0"
+    lines = [
+        "Data type: foo; Version: 1\n"
+        "head1, head2, head 3\n",
+        "Head 1, Head 2, Head 3\n",
+        "1, 2, 3\n",
+        "\n",
+        "1, 2, 3\n",
+    ]
+    _xsv_parse_fail(temp_dir, lines, parse_csv, err)
+
+
+def test_xsv_parse_fail_duplicate_headers(temp_dir: Path):
+    err = "Duplicate column name: head3"
```

Comment:
👍

Reply:
Do we need to test with duplicates on both lines?

Reply:
I do right? it's the second assert for this method. If that's what not you mean can you give an example?

Reply:
Yes you do, didn't delete that comment before the end of the review.